chain/chain/src/resharding.rs

```diff
@@ -66,23 +91,58 @@ fn apply_delayed_receipts<'a>(
     Ok(new_state_roots)
 }
 
+// function to set up flat storage status to Ready after a resharding event
+// TODO(shreyan) : Consolidate this with setting up flat storage during state sync logic
+fn set_flat_storage_state(
+    store: Store,
+    flat_storage_manager: &FlatStorageManager,
+    shard_uid: ShardUId,
+    block_info: BlockInfo,
+) -> Result<(), Error> {
+    assert!(flat_storage_manager.get_flat_storage_for_shard(shard_uid).is_none());
+
+    let mut store_update = store.store_update();
+    store_helper::set_flat_storage_status(
+        &mut store_update,
+        shard_uid,
+        FlatStorageStatus::Ready(FlatStorageReadyStatus { flat_head: block_info }),
+    );
+    store_update.commit()?;
+    flat_storage_manager.create_flat_storage_for_shard(shard_uid)?;
+    Ok(())
+}
+
 impl Chain {
     pub fn build_state_for_split_shards_preprocessing(
         &self,
         sync_hash: &CryptoHash,
         shard_id: ShardId,
         state_split_scheduler: &dyn Fn(StateSplitRequest),
     ) -> Result<(), Error> {
-        let (epoch_id, next_epoch_id) = {
-            let block_header = self.get_block_header(sync_hash)?;
-            (block_header.epoch_id().clone(), block_header.next_epoch_id().clone())
-        };
-        let shard_layout = self.epoch_manager.get_shard_layout(&epoch_id)?;
-        let next_epoch_shard_layout = self.epoch_manager.get_shard_layout(&next_epoch_id)?;
+        let block_header = self.get_block_header(sync_hash)?;
+        let shard_layout = self.epoch_manager.get_shard_layout(block_header.epoch_id())?;
+        let next_epoch_shard_layout =
+            self.epoch_manager.get_shard_layout(block_header.next_epoch_id())?;
+        assert_ne!(shard_layout, next_epoch_shard_layout);
+
         let shard_uid = ShardUId::from_shard_id_and_layout(shard_id, &shard_layout);
-        let prev_hash = *self.get_block_header(sync_hash)?.prev_hash();
+        let prev_hash = *block_header.prev_hash();
         let state_root = *self.get_chunk_extra(&prev_hash, &shard_uid)?.state_root();
-        assert_ne!(shard_layout, next_epoch_shard_layout);
+
+        if let Some(flat_storage_manager) = self.runtime_adapter.get_flat_storage_manager() {
+            for shard_uid in next_epoch_shard_layout
+                .get_split_shard_uids(shard_id)
+                .expect("child shard uids must exist")
+            {
+                let store = self.runtime_adapter.store().clone();
+                let block_info = BlockInfo {
```

Comment:
TODO: Check if we are setting the head correctly

Reply:
We need to set prev_block_info here. Verified via implementation in a separate branch.